### PR TITLE
Noesis template GUI

### PR DESCRIPTION
**Pull request by norda97**
Finishes tasks
 - Design Template GUI
 - Create Main Menu

Created main menu state in CrazyCanvas which instantiates a XAML file menu with buttons to access PlaySesssionState and Benchmark State.

Misc:
 - Fixed BUG where STB was not created and Raytracing render stage tried to execute.
 - Added sleeping functionality to rendergraph for stages

### CODE REVIEW

File: CrazyCanvas/Source/GUI/MainMenuGUI.cpp [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+	m_RayTracingEnabled = EngineConfig::GetBoolProperty("RayTracingEnabled");
+}
+
+MainMenuGUI::~MainMenuGUI()
+{
+}
+
+
+
+bool MainMenuGUI::ConnectEvent(Noesis::BaseComponent* source, const char* event, const char* handler)
```
Comment (skumbananaaa) [inline CrazyCanvas/Source/GUI/MainMenuGUI.cpp]: pSource, pEvent, pHandler

File: CrazyCanvas/Source/States/MainMenuState.cpp [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+#include "Rendering/RenderGraphSerializer.h"
+#include "Rendering/RenderAPI.h"
+
+#include "Engine/EngineConfig.h"
+#include "Game/ECS/Systems/Rendering/RenderSystem.h"
+
+
+MainMenuState::~MainMenuState()
+{
+	int32 ref = m_MainMenuGUI->GetNumReferences();
```
Comment (skumbananaaa) [inline CrazyCanvas/Source/States/MainMenuState.cpp]: Ta bort denna

File: CrazyCanvas/engine_config.json [hunk 3]
```diff
@@ -4,14 +4,11 @@
     1080
   ],
   "FixedTimestep": 60,
-  "RayTracingEnabled": true,
+  "RayTracingEnabled": false,
   "MeshShadersEnabled": false,
```
Comment (skumbananaaa) [inline CrazyCanvas/engine_config.json]: Enable

**Review by skumbananaaa — COMMENTED** [review]
Some Stuff

File: LambdaEngine/Source/GUI/Core/GUIRenderer.cpp [hunk 4]
```diff
@@ -548,78 +548,85 @@ namespace LambdaEngine
...
+			m_ModFrameIndex = modFrameIndex;
+			m_BackBufferIndex = backBufferIndex;
 
-			if (!frameBuffersToRemove.IsEmpty())
+			//Delete Buffers
 			{
-				for (Buffer* pBuffer : frameBuffersToRemove)
+				TArray<Buffer*>& frameBuffersToRemove = m_pBuffersToRemove[m_ModFrameIndex];
+
+				if (!frameBuffersToRemove.IsEmpty())
```
Comment (TheoBerlin) [inline LambdaEngine/Source/GUI/Core/GUIRenderer.cpp]: Unnecessary check

File: LambdaEngine/Source/GUI/Core/GUIRenderer.cpp [hunk 5]
```diff
@@ -548,78 +548,85 @@ namespace LambdaEngine
...
-		{
-			TArray<Buffer*>& buffersNowAvailable = m_pUsedParamsBuffers[m_ModFrameIndex];
 
-			if (!buffersNowAvailable.IsEmpty())
+			//Make Param Buffers available again
 			{
-				for (Buffer* pParamsBuffer : buffersNowAvailable)
+				TArray<Buffer*>& buffersNowAvailable = m_pUsedParamsBuffers[m_ModFrameIndex];
+
+				if (!buffersNowAvailable.IsEmpty())
```
Comment (TheoBerlin) [inline LambdaEngine/Source/GUI/Core/GUIRenderer.cpp]: Same as above, unnecessary

File: LambdaEngine/Source/GUI/Core/GUIRenderer.cpp [hunk 6]
```diff
@@ -548,78 +548,85 @@ namespace LambdaEngine
...
-		{
-			TArray<DescriptorSet*>& descriptorsNowAvailable = m_pUsedDescriptorSets[m_ModFrameIndex];
-
-			if (!descriptorsNowAvailable.IsEmpty())
+			//Make Descriptor Sets available again
 			{
-				for (DescriptorSet* pDescriptorSet : descriptorsNowAvailable)
+				TArray<DescriptorSet*>& descriptorsNowAvailable = m_pUsedDescriptorSets[m_ModFrameIndex];
+
+				if (!descriptorsNowAvailable.IsEmpty())
```
Comment (TheoBerlin) [inline LambdaEngine/Source/GUI/Core/GUIRenderer.cpp:587]: Same as above, unnecessary

File: LambdaEngine/Source/Rendering/RenderGraph.cpp [hunk 7]
```diff
@@ -438,6 +438,22 @@ namespace LambdaEngine
...
+
+		if (it != m_RenderStageMap.end())
+		{
+			RenderStage* pRenderStage = &m_pRenderStages[it->second];
+			pRenderStage->Sleeping = sleeping;
+		}
+		else
+		{
+			LOG_WARNING("[RenderGraph]: SetRenderStageSleeping failed, render stage with name \"%s\" could not be found", renderStageName.c_str());
+			return;
```
Comment (TheoBerlin) [inline LambdaEngine/Source/Rendering/RenderGraph.cpp:453]: Unnecessary return

File: LambdaEngine/Source/Rendering/RenderGraph.cpp [hunk 8]
```diff
@@ -3604,7 +3621,7 @@ namespace LambdaEngine
 		CommandList*		pGraphicsCommandList,
 		CommandList**		ppExecutionStage)
 	{
-		if (pRenderStage->FrameCounter != pRenderStage->FrameOffset && pRenderStage->pDisabledRenderPass == nullptr)
+		if ((pRenderStage->FrameCounter != pRenderStage->FrameOffset || pRenderStage->Sleeping) && pRenderStage->pDisabledRenderPass == nullptr )
```
Comment (TheoBerlin) [inline LambdaEngine/Source/Rendering/RenderGraph.cpp:3638]:
```suggestion
		if ((pRenderStage->FrameCounter != pRenderStage->FrameOffset || pRenderStage->Sleeping) && pRenderStage->pDisabledRenderPass == nullptr)
```

File: LambdaEngine/Source/Rendering/RenderGraph.cpp [hunk 9]
```diff
@@ -3715,8 +3732,8 @@ namespace LambdaEngine
 
 				clearColorCount++;
 			}
-
-			if (pRenderStage->FrameCounter == pRenderStage->FrameOffset)
+
```
Comment (TheoBerlin) [inline LambdaEngine/Source/Rendering/RenderGraph.cpp]:
```suggestion

```

File: Sandbox/Source/States/SandboxState.cpp [hunk 10]
```diff
@@ -51,8 +51,14 @@ SandboxState::SandboxState(LambdaEngine::State* pOther) : LambdaEngine::State(pO
 
 SandboxState::~SandboxState()
 {
-	m_GUITest.Reset();
-	m_View.Reset();
+	if (m_GUITest.GetPtr() != nullptr)
+	{
+		int32 ref = m_GUITest->GetNumReferences();
+
```
Comment (frozenbanana) [inline Sandbox/Source/States/SandboxState.cpp:60]: Looks like it is not needed.

Comment (norda97) [reply Sandbox/Source/States/SandboxState.cpp:60]: Already fixed

File: LambdaEngine/Source/Rendering/LineRenderer.cpp [hunk 11]
```diff
@@ -358,9 +358,13 @@ namespace LambdaEngine
 		uint32 modFrameIndex,
 		uint32 backBufferIndex,
 		CommandList** ppFirstExecutionStage,
-		CommandList** ppSecondaryExecutionStage)
+		CommandList** ppSecondaryExecutionStage,
+		bool Sleeping)
```
Comment (frozenbanana) [inline LambdaEngine/Source/Rendering/LineRenderer.cpp:362]: Functions use pascalcase. Parameters and local variables use camelcase. `bool Sleeping` -> `bool sleeping`

File: LambdaEngine/Source/Rendering/ImGuiRenderer.cpp [hunk 12]
```diff
@@ -302,10 +302,14 @@ namespace LambdaEngine
 		uint32 modFrameIndex,
 		uint32 backBufferIndex,
 		CommandList** ppFirstExecutionStage,
-		CommandList** ppSecondaryExecutionStage)
+		CommandList** ppSecondaryExecutionStage,
+		bool Sleeping)
```
Comment (frozenbanana) [inline LambdaEngine/Source/Rendering/ImGuiRenderer.cpp]: Functions use pascalcase. Parameters and local variables use camelcase. `bool Sleeping` -> `bool sleeping`

Comment (norda97) [reply LambdaEngine/Source/Rendering/ImGuiRenderer.cpp]: Done

File: CrazyCanvas/Source/GUI/MainMenuGUI.cpp [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+void MainMenuGUI::OnButton1Click(Noesis::BaseComponent* pSender, const Noesis::RoutedEventArgs& args)
+{
+	UNREFERENCED_VARIABLE(pSender);
+	UNREFERENCED_VARIABLE(args);
+
+	using namespace LambdaEngine;
+	SetRenderStagesSleeping();
+
+	State* pStartingState = DBG_NEW PlaySessionState();
+	StateManager::GetInstance()->EnqueueStateTransition(pStartingState, STATE_TRANSITION::PUSH);
```
Comment (TheoBerlin) [inline CrazyCanvas/Source/GUI/MainMenuGUI.cpp]:
```suggestion
	StateManager::GetInstance()->EnqueueStateTransition(pStartingState, STATE_TRANSITION::POP_AND_PUSH);
```

File: CrazyCanvas/Source/GUI/MainMenuGUI.cpp [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+
+
+
+bool MainMenuGUI::ConnectEvent(Noesis::BaseComponent* source, const char* event, const char* handler)
+{
+	NS_CONNECT_EVENT(Noesis::Button, Click, OnButton1Click);
+	NS_CONNECT_EVENT(Noesis::Button, Click, OnButton2Click);
+	NS_CONNECT_EVENT(Noesis::CheckBox, Click, OnRayTracingChecked);
+	NS_CONNECT_EVENT(Noesis::CheckBox, Click, OnMeshShadersChecked);
+	return false;
```
Comment (TheoBerlin) [inline CrazyCanvas/Source/GUI/MainMenuGUI.cpp:37]: Why does it always return false?

**Review by TheoBerlin — CHANGES_REQUESTED** [review]

File: CrazyCanvas/Include/States/MainMenuState.h [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+	void Resume() override final {};
+	void Pause() override final {};
+
+	void Tick(LambdaEngine::Timestamp delta) override final;
+
+
+private:
+
+	Noesis::Ptr<MainMenuGUI> m_MainMenuGUI;
+	Noesis::Ptr<Noesis::IView> m_View;
```
Comment (Vonsen) [inline CrazyCanvas/Include/States/MainMenuState.h:29]:
```suggestion
	Noesis::Ptr<MainMenuGUI> m_pMainMenuGUI;
	Noesis::Ptr<Noesis::IView> m_pView;
```

File: CrazyCanvas/Source/GUI/MainMenuGUI.cpp (hunk 13 above)
Comment (norda97) [reply CrazyCanvas/Source/GUI/MainMenuGUI.cpp]: Done

File: CrazyCanvas/Source/GUI/MainMenuGUI.cpp (hunk 14 above)
Comment (norda97) [reply CrazyCanvas/Source/GUI/MainMenuGUI.cpp:37]: It should be false here

**Review by TheoBerlin — APPROVED** [review]

**Pull request merged**